### ISSUE TITLE
failure on Windows

### ISSUE THREAD

**Issue by maelle** [ropensci/vcr#158]
cf https://github.com/ropensci/ropenaq/runs/469919162?check_suite_focus=true

I'm guessing it's an encoding issue :wink: :sob:

**Comment by sckott** [comment]
did you try setting `preserve_exact_body_bytes = TRUE` yet? it may be that yaml pkg is not handling the characters well, but might be fine if its a base64 encoded string

**Comment by maelle** [comment]
Thanks! I get new failures now https://github.com/ropensci/ropenaq/runs/470099508

**Comment by sckott** [comment]
it looks like you have the same cassettes though - i should have said use `preserve_exact_body_bytes` and re-record cassettes

**Comment by maelle** [comment]
I thought I had https://github.com/ropensci/ropenaq/commit/3be1989ac607f604a1b02e8767c2208a4373620d but I'll retry

**Comment by maelle** [comment]
This is my config

```r
library("vcr")
invisible(vcr::vcr_configure(
  dir = "../fixtures",
  preserve_exact_body_bytes = TRUE
))
```

Re-recording doesn't seem to change much https://github.com/ropensci/ropenaq/commit/d68e840000c2de917505ade4d91175f3eda245aa

**Comment by sckott** [comment]
what version of `vcr` do you have?

**Comment by maelle** [comment]
0.4.0, I'll upgrade to the dev version!

**Comment by maelle** [comment]
Ok, I updated webmockr and vcr using their dev version (and set the `Remotes` field of `DESCRIPTION` to point to them), recorded the cassettes now I get only one failure in one test. The problem is no longer encoding in the body, but in the URI. :grimacing: 

https://github.com/ropensci/ropenaq/runs/471886535?check_suite_focus=true

```r
An HTTP request has been made that vcr does not know how to handle:
GET https://api.openaq.org/v1/measurements?country=DE&city=J%C3%AF%C2%BF%C2%BDrgen%20Friesel&limit=1&page=1
vcr is currently using the following cassette:
  - ../fixtures/buildQueries_accents.yml
    - record_mode: once
    - match_requests_on: method, uri
```

It's a test called "Queries work with spaces and accents". [Cassette](https://github.com/ropensci/ropenaq/blob/f8453baf740587de33f3a2afcf21fa4ff44ae382/tests/fixtures/buildQueries_accents.yml#L537);  [test](https://github.com/ropensci/ropenaq/blob/f8453baf740587de33f3a2afcf21fa4ff44ae382/tests/testthat/test-buildQueries.R#L84).

**Comment by sckott** [comment]
thanks, I'll have a look

**Comment by sckott** [comment]
works fine locally. ugh, i guess an encoding issue on windows only. imagine it's in the code for comparing requests.

**Comment by sckott** [comment]
I was able to replicate the issue on a windows VM - I can't figure out what the non-escaped query string should look like for the failing test. I get 

```r
curl::curl_unescape("J%EF%BF%BDrgen+Friesel")
#> [1] "J�rgen+Friesel"
```

to clarify (and still on windows): I could only replicate the error when using your original casette file in your github repo. Hoewever, if I deleted the cassette file, then ran tests again, it did not fail. So perhaps the cassette is recorded on a different operating system than where the error is happening, and somehow the encoding is handled differently

**Comment by maelle** [comment]
> So perhaps the cassette is recorded on a different operating system than where the error is happening, and somehow the encoding is handled differently

So how should I proceed to try and help debug this next week?

**Comment by sckott** [comment]
we do a pretty simple comparison of the URIs here https://github.com/ropensci/vcr/blob/master/R/request_matcher_registry.R#L48 - i guess we need to figure out a way to make sure the two strings are not messed up on windows, not sure yet

**Comment by maelle** [comment]
maybe the two URIs should be compared after being `URLencode()`ed or so?

**Comment by maelle** [comment]
New platform to reproduce the error :smirk: https://cloud.r-project.org/web/checks/check_results_ropenaq.html i.e. https://blog.r-hub.io/2019/04/25/r-devel-linux-x86-64-debian-clang/#r-devel-linux-x86-64-debian-clang

**Comment by sckott** [comment]
thanks, okay, but we can't test `r-devel-linux-x86_64-debian-clang` on rhub right now though right? rub bug or something

**Comment by maelle** [comment]
Right. That said any R-hub Windows platform should reproduce the bug.

**Comment by sckott** [comment]
Sorry for the delay on this @maelle - I assume this is still a problem, yes?

**Comment by maelle** [comment]
I think so but haven't looked into it for a long time (thankfully you know I'll soon spend more time thinking about http testing!)

The test was https://github.com/ropensci/ropenaq/blob/ecd9d8635f761811edc99968609c3ff3ae62b21f/tests/testthat/test-buildQueries.R#L83

I see another test is failing at the moment, unrelated to vcr, sorry about that.

**Comment by maelle** [comment]
new sighting https://github.com/rladies/meetupr/runs/1435609920?check_suite_focus=true
files https://github.com/rladies/meetupr/pull/84/files / the branch is "noninteractive"

**Comment by maelle** [comment]
I tried adding `URLencode()` at https://github.com/ropensci/vcr/blob/master/R/request_matcher_registry.R?rgh-link-date=2020-02-28T17%3A56%3A12Z#L48, it didn't work.

It'd be very nice to fix this. :thinking:

**Comment by maelle** [comment]
Could it work to encode the uri at the time of serializing it to YAML?

**Comment by maelle** [comment]
With JSON the failure is different.

**Comment by sckott** [comment]
Did this get sorted yet?

**Comment by maelle** [comment]
I don't think so, do you need write access to https://github.com/maelle/vcrencoding?

**Comment by yogat3ch** [comment]
Hi @sckott & @maelle,
It looks like I'm also encountering this issue on a Windows 10 machine - it appears to be URIs initially sent as un-encoded URIs are URL encoded in the yml files. When a test is run again, `vcr` is not matching the un-encoded request URI with the URIs in the yml files that are encoded.
Do we need another reprex of this behavior or does @maelle's reprex work?

**Comment by yogat3ch** [comment]
@sckott Ok, finally figured out how to make a reprex for this behavior:
```
.url <- httr::parse_url("https://httpbin.org/get")
`%>%` <- dplyr::`%>%`

.url$query <- lubridate::ymd_hms("2021-06-11 14:28:13") %>% 
  list(start = ., end = . - 1) %>% 
  purrr::map(~stringr::str_replace(format(.x, "%Y-%m-%dT%H:%M:%S%z"), "[\\+\\-](\\d{2})(\\d{2})$", "-\\1:\\2")) %>% 
  append(list(limit = 10000, timeframe = "1Min"))
.t <- .url$query$start
.url <- httr::build_url(.url)
vcr::vcr_configure(write_disk_path = ".", dir = ".")
vcr::use_cassette("test1", {
  (resp <- httr::GET(utils::URLdecode(.url)))
  testthat::test_that("blah", {
    testthat::expect_true(httr::content(resp)$args$start == .t)
  })  
})
```
It definitely has to do with the URL encoding because if `utils::URLdecode` is removed it works fine.

**Comment by sckott** [comment]
thanks @yogat3ch will have another look at this
